### PR TITLE
Datatype Terms

### PR DESCRIPTION
**Pull request by afnaanhashmi**

### CODE REVIEW

File: src/generic_sort.cpp [hunk 1]
```diff
@@ -373,4 +385,52 @@ std::string GenericDatatypeSort::to_string() const
...
 }
 
+DatatypeComponentSort::DatatypeComponentSort(SortKind sk,
+                                             std::string name,
+                                             Sort dt)
+    : GenericSort(sk), name(name), dt_sort(dt)
+{
+  if (sk != CONSTRUCTOR && sk != SELECTOR && sk != TESTER)
+  {
+    throw "Wrong sortkind input";
```
Comment (4tXJ7f) [inline src/generic_sort.cpp:395]: It seems like we usually throw `IncorrectUsageException` for those cases:

```suggestion
    throw IncorrectUsageException("Wrong sortkind input");
```

File: src/ops.cpp [hunk 2]
```diff
@@ -91,7 +91,7 @@ const std::unordered_map<PrimOp, std::string> primop2str(
       { Exists, "exists" },
       { Apply_Selector, "apply_selector" },
       { Apply_Tester, "apply_tester" },
-      { Apply_Constructor, "apply_constructor" } });
+      { Apply_Constructor, "apply_selector" } });
```
Comment (4tXJ7f) [inline src/ops.cpp:94]: This is a bit surprising. Is it supposed to be `apply_selector`?

File: src/sort_inference.cpp [hunk 3]
```diff
@@ -536,4 +564,21 @@ Sort store_sort(Op op, const AbsSmtSolver * solver, const SortVec & sorts)
   return arraysort;
 }
 
+Sort selector_sort(Op op, const AbsSmtSolver * solver, const SortVec & sorts)
+{
+  Sort parent_sort = (sorts[0])->get_domain_sorts()[0];
+  // shared_ptr<GenericDatatype> dt =
+  // static_pointer_cast<GenericDatatype>(parent_sort->get_datatype());
```
Comment (4tXJ7f) [inline src/sort_inference.cpp:571]:
```suggestion
```

File: src/generic_solver.cpp [hunk 4]
```diff
@@ -605,18 +641,95 @@ void GenericSolver::add_selector_self(DatatypeConstructorDecl & dt, const std::s
...
+        }
+      }
+    }
+  }
+  if (!found)
+  {
+    throw InternalSolverException("Selector not in datatype");
+  }
+
+  // Sort cons_sort = make_generic_sort(SELECTOR, name, s);
```
Comment (4tXJ7f) [inline src/generic_solver.cpp:727]:
```suggestion
```

File: src/generic_solver.cpp [hunk 5]
```diff
@@ -357,7 +382,18 @@ std::string GenericSolver::to_smtlib_def(Term term) const
...
       }
     }
-    result += ")";
+    if (gt->get_op() == Apply_Constructor)
+    {
+      shared_ptr<GenericDatatype> dt = static_pointer_cast<GenericDatatype>(
+          (gt->get_sort())->get_datatype());
+      result += dt->get_num_selectors((*term_name_map)[gt->get_children()[0]])
+                    ? ")"
+                    : "";
```
Comment (4tXJ7f) [inline src/generic_solver.cpp:391]: Instead of repeating the check here, it would be a bit better to define some boolean, e.g., `nullary_constructor` and only perform the check once.

File: src/generic_sort.cpp [hunk 6]
```diff
@@ -187,6 +191,13 @@ string GenericSort::compute_string() const {
     {
       return static_pointer_cast<GenericDatatype>(get_datatype())->get_name();
     }
+    else if (get_sort_kind() == SortKind::CONSTRUCTOR
+             || get_sort_kind() == SortKind::SELECTOR
+             || get_sort_kind() == SortKind::TESTER)
+    {
+      cout << "in comptue" << endl;
```
Comment (4tXJ7f) [inline src/generic_sort.cpp:198]:
```suggestion
```

File: src/generic_sort.cpp [hunk 7]
```diff
@@ -196,6 +207,7 @@ string GenericSort::compute_string() const {
 
 SortKind GenericSort::get_sort_kind() const { return sk; }
 
+
```
Comment (4tXJ7f) [inline src/generic_sort.cpp:210]:
```suggestion
```

File: src/generic_sort.cpp [hunk 8]
```diff
@@ -17,7 +17,6 @@
 
 #include <functional>
 #include <unordered_map>
-
```
Comment (4tXJ7f) [inline src/generic_sort.cpp:20]: Was this change done by ClangFormat? If not, I'd prefer to exclude it from this PR.

File: tests/test-smtlib-reader.cpp [hunk 9]
```diff
@@ -18,6 +18,7 @@
 #define STRFY(A) STRHELPER(A)
 
 #include <gtest/gtest.h>
+
```
Comment (4tXJ7f) [inline tests/test-smtlib-reader.cpp:21]: It is better not to include changes in unrelated files in a PR.

```suggestion
```

File: src/sort_inference.cpp [hunk 10]
```diff
@@ -368,6 +375,27 @@ bool check_select_sorts(const SortVec & sorts)
...
+  {
+    return false;
+  }
+
+  Sort dt_sort = sorts[0];
+  if (dt_sort->get_sort_kind() != DATATYPE)
+  {
+    return false;
+  }
+  return true;
```
Comment (4tXJ7f) [inline src/sort_inference.cpp:391]:
```suggestion
  return dt_sort->get_sort_kind() == DATATYPE;
```

File: include/sorting_network.h [hunk 11]
```diff
@@ -91,7 +91,6 @@ class SortingNetwork
    *  @return the combined vector
    */
   TermVec merge(const TermVec & sorted1, const TermVec & sorted2) const;
-
```
Comment (4tXJ7f) [inline include/sorting_network.h:94]: It would be better not to include this change in this PR since it is in an unrelated file.

File: cvc4/src/cvc4_sort.cpp [hunk 12]
```diff
@@ -152,6 +152,10 @@ SortKind CVC4Sort::get_sort_kind() const
   {
     return DATATYPE;
   }
+  else if (sort.isConstructor())
+  {
+    return CONSTRUCTOR;
+  }
```
Comment (4tXJ7f) [inline cvc4/src/cvc4_sort.cpp:158]: What about `SELECTOR` and `TESTER`? It looks like we have the necessary functionality in the CVC4/cvc5 [API](https://github.com/cvc5/cvc5/blob/3dda54ba7e6952060766775c56969ab920430a8a/src/api/cvc4cpp.h#L386-L396).

File: src/generic_datatype.cpp [hunk 13]
```diff
@@ -135,20 +135,25 @@ int GenericDatatype::get_num_selectors(std::string cons) const
...
   {
     if (static_pointer_cast<GenericDatatypeConstructorDecl>(cons_decl_vector[i])
             ->get_name()
         == cons)
     {
+      found = true;
       // Calls the constructor's get_selector_count() function
       num_selectors = static_pointer_cast<GenericDatatypeConstructorDecl>(
                           cons_decl_vector[i])
                           ->get_selector_count();
```
Comment (4tXJ7f) [inline src/generic_datatype.cpp:150]: This is somewhat unrelated to the PR but can't we `break` once we found the constructor?

File: include/generic_sort.h [hunk 14]
```diff
@@ -207,4 +209,24 @@ class GenericDatatypeSort : public GenericSort
   Datatype gdt;
 };
 
+class DatatypeComponentSort : public GenericSort
```
Comment (4tXJ7f) [inline include/generic_sort.h:212]: It would be good to add a comment about the purpose of this sort (most of the classes in this file don't have a comment but it would be good to have regardless).

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
This is great, thanks for doing all of this! Some quick comments. I'll take a closer look later. Please let me know if anything is unclear.

File: tests/test-smtlib-reader.cpp [hunk 15]
```diff
@@ -112,63 +100,6 @@ TEST_P(IntReaderTests, QF_UFLIA_Smt2Files)
   }
 }
 
-TEST_P(BitVecReaderTests, QF_UFBV_Smt2Files)
```
Comment (4tXJ7f) [inline tests/test-smtlib-reader.cpp:115]: There are some removals in this file that seem unrelated to the PR. Are those on purpose?

File: src/generic_sort.cpp [hunk 16]
```diff
@@ -373,4 +382,52 @@ std::string GenericDatatypeSort::to_string() const
...
+{
+  return name;
+}
+
+SortVec DatatypeComponentSort::get_domain_sorts() const
+{
+  std::vector<Sort> domain_sorts;
+  domain_sorts.push_back(dt_sort);
+  return domain_sorts;
+}
```
Comment (4tXJ7f) [inline src/generic_sort.cpp:412]: I am not sure if having the datatype as the domain sort is the right thing to do here. 
Instead, I would change `DatatypeComponentSort` to use domain/co-domain as follows :

- Constructor: domain = arguments passed to the constructor, co-domain = datatype
- Tester: domain = datatype, co-domain = bool
- Selector: domain = datatype, co-domain = sort of the selected parameter

What do you think?

File: tests/test-smtlib-reader.cpp (hunk 15 above)
Comment (afnaanhashmi) [reply tests/test-smtlib-reader.cpp:115]: No those are not on purpose. I don't remember making any changes to that file, and tried at least to remove it from the pr.

File: src/generic_solver.cpp [hunk 17]
```diff
@@ -605,18 +638,93 @@ void GenericSolver::add_selector_self(DatatypeConstructorDecl & dt, const std::s
...
+      static_pointer_cast<GenericDatatype>(s->get_datatype());
+  bool found = false;
+  for (int i = 0; i < dt->get_num_constructors(); ++i)
+  {
+    if (static_pointer_cast<GenericDatatypeConstructorDecl>(
+            dt->get_cons_vector()[i])
+            ->get_name()
+        == name)
+    {
+      found = true;
```
Comment (4tXJ7f) [inline src/generic_solver.cpp:679]:
```suggestion
      found = true;
      break;
```

File: src/generic_solver.cpp [hunk 18]
```diff
@@ -605,18 +638,93 @@ void GenericSolver::add_selector_self(DatatypeConstructorDecl & dt, const std::s
...
+            dt->get_cons_vector()[i]);
+    if (curr_con->get_name() == con)
+    {
+      for (int f = 0; f < curr_con->get_selector_count(); ++f)
+      {
+        if (((curr_con->get_selector_vector())[f]).name == name)
+        {
+          found = true;
+          static_pointer_cast<DatatypeComponentSort>(cons_sort)
+              ->set_selector_sort(((curr_con->get_selector_vector())[f]).sort);
```
Comment (4tXJ7f) [inline src/generic_solver.cpp:714]:
```suggestion
          found = true;
          static_pointer_cast<DatatypeComponentSort>(cons_sort)
              ->set_selector_sort(((curr_con->get_selector_vector())[f]).sort);
          break;
```

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Just some minor suggestions. Looking good otherwise :)

File: src/sort_inference.cpp [hunk 19]
```diff
@@ -536,4 +578,19 @@ Sort store_sort(Op op, const AbsSmtSolver * solver, const SortVec & sorts)
   return arraysort;
 }
 
+Sort selector_sort(Op op, const AbsSmtSolver * solver, const SortVec & sorts)
+{
+  Sort parent_sort = (sorts[0])->get_domain_sorts()[0];
+  return static_pointer_cast<DatatypeComponentSort>(sorts[0])
+      ->get_selector_sort();
```
Comment (4tXJ7f) [inline src/sort_inference.cpp:585]:
```suggestion
  return static_pointer_cast<DatatypeComponentSort>(sorts[0])
      ->get_codomain_sort();
```

File: include/generic_sort.h [hunk 20]
```diff
@@ -207,4 +208,29 @@ class GenericDatatypeSort : public GenericSort
...
+{
+ public:
+  DatatypeComponentSort(SortKind sk, std::string name, Sort dt);
+  ~DatatypeComponentSort(){};
+  std::string compute_string() const override;
+  std::string to_string() const override;
+  std::string get_uninterpreted_name() const override;
+  SortVec get_domain_sorts() const override;
+  Sort get_codomain_sort() const override;
+  Sort get_selector_sort() const;
```
Comment (4tXJ7f) [inline include/generic_sort.h:225]:
```suggestion
```

File: src/generic_sort.cpp [hunk 21]
```diff
@@ -373,4 +382,93 @@ std::string GenericDatatypeSort::to_string() const
...
+  else if (sk == SELECTOR)
+  {
+    return selector_sort;
+  }
+
+  assert (sk == CONSTRUCTOR || sk == TESTER || sk == SELECTOR);
+}
+
+Sort DatatypeComponentSort::get_selector_sort() const { return selector_sort; }
+
```
Comment (4tXJ7f) [inline src/generic_sort.cpp:456]:
```suggestion
```

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Just three more very minor suggestions. Could you also remove the changes in `tests/test-smtlib-reader.cpp`? I think that we can go ahead and merge afterwards :)

**Review by 4tXJ7f — APPROVED** [review]
LGTM, thanks for all the changes!
